README
NAME
    Log::FreeSWITCH::Line - FreeSWITCH log line parsing and serializing.

SYNOPSIS
     use Log::FreeSWITCH::Line qw(parse serialize);

     my $data_o = parse($data);
     my $data = serialize($data_o);

SUBROUTINES
  "parse"
     my $data_o = parse($data);

    Parse FreeSWITCH log line.

    Returns Log::FreeSWITCH::Line::Data object.

  "serialize"
     my $data = serialize($data_o);

    Serialize Log::FreeSWITCH::Line::Data object to FreeSWITCH log line.

    Returns string.

ERRORS
     parse():
             Cannot parse data.
                     Data: %s

     serialize():
             Serialize object must be 'Log::FreeSWITCH::Line::Data' object.

EXAMPLE1
     use strict;
     use warnings;

     use Data::Printer;
     use Log::FreeSWITCH::Line qw(parse);

     # Log record.
     my $data = '2014-07-01 13:37:53.973562 [DEBUG] sofia.c:4045 inbound-codec-prefs [PCMA]';

     # Parse.
     my $data_o = parse($data);

     # Dump.
     p $data_o;

     # Output:
     # Log::FreeSWITCH::Line::Data  {
     #     Parents       Mo::Object
     #     public methods (0)
     #     private methods (1) : _datetime
     #     internals: {
     #         date        "2014-07-01",
     #         file        "sofia.c",
     #         file_line   4045,
     #         message     "inbound-codec-prefs [PCMA]",
     #         raw         "2014-07-01 13:37:53.973562 [DEBUG] sofia.c:4045 inbound-codec-prefs [PCMA]",
     #         time        "13:37:53.973562",
     #         type        "DEBUG"
     #     }
     # }

EXAMPLE2
     use strict;
     use warnings;

     use Log::FreeSWITCH::Line qw(serialize);
     use Log::FreeSWITCH::Line::Data;

     # Data.
     my $record = Log::FreeSWITCH::Line::Data->new(
             'date' => '2014-07-01',
             'file' => 'sofia.c',
             'file_line' => 4045,
             'message' => 'inbound-codec-prefs [PCMA]',
             'time' => '13:37:53.973562',
             'type' => 'DEBUG',
     );

     # Serialize and print to stdout.
     print serialize($record)."\n";

     # Output:
     # 2014-07-01 13:37:53.973562 [DEBUG] sofia.c:4045 inbound-codec-prefs [PCMA]

DEPENDENCIES
    Error::Pure, Exporter, Log::FreeSWITCH::Line::Data, Readonly,
    Scalar::Util.

SEE ALSO
    Log::FreeSWITCH::Line::Data
        Data object which represents FreeSWITCH log line.

REPOSITORY
    <https://github.com/michal-josef-spacek/Log-FreeSWITCH-Line>

AUTHOR
    Michal Josef Špaček <mailto:skim@cpan.org>

    <http://skim.cz>

LICENSE AND COPYRIGHT
    © 2014-2023 Michal Josef Špaček

    BSD 2-Clause License

VERSION
    0.09



Changes
0.09
 [Common]
 - Update copyright years.
 - Update Module::Install to 1.21 version.

0.08 2022-12-09T20:11:29+01:00
 [Common]
 - Rename example files to better names.
 - Rewrite doc to use =head2 sections.
 - Update copyright years.

 [Log::FreeSWITCH::Line::Data]
 - Add test with bad date.

0.07 2021-02-21T15:39:45+01:00
 [Common]
 - Fix bugtracker.
 - Fix use lib '.' in Makefile.PL.
 - Improve LICENSE AND COPYRIGHT section in doc.
 - Improve SYNOPSIS section in doc.
 - Update Module::Install to 1.19 version.
 - Update author github username.
 - Update author name.
 - Update copyright years.

0.06 2015-10-22T11:27:41+02:00
 [Common]
 - Move author tests to xt/ directory.

 [Log::FreeSWITCH::Line]
 - Add SEE ALSO section to doc.

 [Log::FreeSWITCH::Line::Data]
 - Improve and fix SEE ALSO section in doc.

0.05 2015-05-19T17:02:04+02:00
 [Common]
 - Update Module::Install to 1.16 version.

 [Log::FreeSWITCH::Line]
 - Add tests.
 - Improve parse() subroutine to parse time without miliseconds.
 - Remove warning in serialize() with data object without message.

0.04 2015-01-09T07:42:44+01:00
 [Log::FreeSWITCH::Line::Data]
 - Fix test dependency for DateTime module.

0.03 2015-01-08T19:59:22+01:00
 [Common]
 - Fix dependencies, which breaks tests (bad Error::Pure::Always).

0.02 2015-01-08T01:30:56+01:00
 [Log::FreeSWITCH::Line::Data]
 - Add missing VERSION.
 - Add tests.
 - Fix time without miliseconds.

0.01 2015-01-04T11:16:34+01:00
 - First version.


Line.pm
package Log::FreeSWITCH::Line;

use base qw(Exporter);
use strict;
use warnings;

use Error::Pure qw(err);
use Log::FreeSWITCH::Line::Data;
use Readonly;
use Scalar::Util qw(blessed);

Readonly::Array our @EXPORT_OK => qw{parse serialize};
Readonly::Scalar our $LOG_REGEXP => qr{(\d{4}-\d{2}-\d{2})\s+(\d{2}:\d{2}:\d{2}\.?\d*)\s+\[(\w+)\]\s+([^:]+):(\d+)\s+(.*)};
Readonly::Scalar our $SPACE => q{ };

our $VERSION = 0.09;

# Parse FreeSWITCH log line.
sub parse {
	my $data = shift;
	my $data_o;
	if ($data =~ m/^$LOG_REGEXP$/ms) {
		$data_o = Log::FreeSWITCH::Line::Data->new(
			'date' => $1,
			'file' => $4,
			'file_line' => $5,
			'message' => $6,
			'raw' => $data,
			'time' => $2,
			'type' => $3,
		);
	} else {
		err 'Cannot parse data.',
			'Data', $data;
	}
	return $data_o;
}

# Serialize Log::FreeSWITCH::Line::Data object to FreeSWITCH log line.
sub serialize {
	my $data_o = shift;

	# Check object.
	if (! blessed($data_o) || ! $data_o->isa('Log::FreeSWITCH::Line::Data')) {
		err "Serialize object must be 'Log::FreeSWITCH::Line::Data' object.";
	}

	# Serialize.
	my $data = $data_o->date.
		$SPACE.$data_o->time.
		$SPACE.'['.$data_o->type.']'.
		$SPACE.$data_o->file.':'.$data_o->file_line.
		$SPACE;
	if (defined $data_o->message) {
		$data .= $data_o->message;
	}
	$data_o->raw($data);
	return $data;
}

1;

__END__

=pod

=encoding utf8

=head1 NAME

Log::FreeSWITCH::Line - FreeSWITCH log line parsing and serializing.

=head1 SYNOPSIS

 use Log::FreeSWITCH::Line qw(parse serialize);

 my $data_o = parse($data);
 my $data = serialize($data_o);

=head1 SUBROUTINES

=head2 C<parse>

 my $data_o = parse($data);

Parse FreeSWITCH log line.

Returns Log::FreeSWITCH::Line::Data object.

=head2 C<serialize>

 my $data = serialize($data_o);

Serialize Log::FreeSWITCH::Line::Data object to FreeSWITCH log line.

Returns string.

=head1 ERRORS

 parse():
         Cannot parse data.
                 Data: %s

 serialize():
         Serialize object must be 'Log::FreeSWITCH::Line::Data' object.

=head1 EXAMPLE1

=for comment filename=parse_log_line.pl

 use strict;
 use warnings;

 use Data::Printer;
 use Log::FreeSWITCH::Line qw(parse);

 # Log record.
 my $data = '2014-07-01 13:37:53.973562 [DEBUG] sofia.c:4045 inbound-codec-prefs [PCMA]';

 # Parse.
 my $data_o = parse($data);

 # Dump.
 p $data_o;

 # Output:
 # Log::FreeSWITCH::Line::Data  {
 #     Parents       Mo::Object
 #     public methods (0)
 #     private methods (1) : _datetime
 #     internals: {
 #         date        "2014-07-01",
 #         file        "sofia.c",
 #         file_line   4045,
 #         message     "inbound-codec-prefs [PCMA]",
 #         raw         "2014-07-01 13:37:53.973562 [DEBUG] sofia.c:4045 inbound-codec-prefs [PCMA]",
 #         time        "13:37:53.973562",
 #         type        "DEBUG"
 #     }
 # }

=head1 EXAMPLE2

=for comment filename=serialize_log_structure.pl

 use strict;
 use warnings;

 use Log::FreeSWITCH::Line qw(serialize);
 use Log::FreeSWITCH::Line::Data;

 # Data.
 my $record = Log::FreeSWITCH::Line::Data->new(
         'date' => '2014-07-01',
         'file' => 'sofia.c',
         'file_line' => 4045,
         'message' => 'inbound-codec-prefs [PCMA]',
         'time' => '13:37:53.973562',
         'type' => 'DEBUG',
 );

 # Serialize and print to stdout.
 print serialize($record)."\n";

 # Output:
 # 2014-07-01 13:37:53.973562 [DEBUG] sofia.c:4045 inbound-codec-prefs [PCMA]

=head1 DEPENDENCIES

L<Error::Pure>,
L<Exporter>,
L<Log::FreeSWITCH::Line::Data>,
L<Readonly>,
L<Scalar::Util>.

=head1 SEE ALSO

=over

=item L<Log::FreeSWITCH::Line::Data>

Data object which represents FreeSWITCH log line.

=back

=head1 REPOSITORY

L<https://github.com/michal-josef-spacek/Log-FreeSWITCH-Line>

=head1 AUTHOR

Michal Josef Špaček L<mailto:skim@cpan.org>

L<http://skim.cz>

=head1 LICENSE AND COPYRIGHT

© 2014-2023 Michal Josef Špaček

BSD 2-Clause License

=head1 VERSION

0.09

=cut


Makefile.PL
use lib '.';
use strict;
use warnings;

use inc::Module::Install;

# Definition.
abstract 'FreeSWITCH log line processing.';
author 'Michal Josef Spacek <skim@cpan.org>';
author_requires 'English' => 0;
author_requires 'File::Object' => 0.07;
author_requires 'Test::More' => 0;
author_requires 'Test::NoWarnings' => 0;
author_requires 'Test::Pod' => 0;
author_requires 'Test::Pod::Coverage' => 0;
license 'bsd';
name 'Log-FreeSWITCH-Line';
readme_from 'Line.pm';
recursive_author_tests('xt');
requires 'Error::Pure' => 0.22;
requires 'Error::Pure::Always' => 0.05;
requires 'Mo' => 0;
requires 'Readonly' => 0;
requires 'Scalar::Util' => 0;
requires 'perl' => '5.8.0';
resources 'bugtracker' => 'https://github.com/michal-josef-spacek/Log-FreeSWITCH-Line/issues';
resources 'homepage' => 'https://github.com/michal-josef-spacek/Log-FreeSWITCH-Line';
resources 'repository' => 'git://github.com/michal-josef-spacek/Log-FreeSWITCH-Line';
test_requires 'DateTime' => 0;
test_requires 'English' => 0;
test_requires 'Error::Pure::Utils' => 0.22;
test_requires 'Test::MockObject' => 0;
test_requires 'Test::More' => 0;
test_requires 'Test::NoWarnings' => 0;
tests_recursive;
version '0.09';

# Run.
WriteAll();


Line/Data.pm
package Log::FreeSWITCH::Line::Data;

use strict;
use warnings;

use English;
use Error::Pure::Always;
use Error::Pure qw(err);
use Mo qw(builder is required);

our $VERSION = 0.09;

has date => (
	'is' => 'ro',
	'required' => 1,
);
has datetime_obj => (
	'is' => 'ro',
	'builder' => '_datetime',
);
has file => (
	'is' => 'ro',
	'required' => 1,
);
has file_line => (
	'is' => 'ro',
	'required' => 1,
);
has message => (
	'is' => 'ro',
);
has raw => (
	'is' => 'rw',
);
has time => (
	'is' => 'ro',
	'required' => 1,
);
has type => (
	'is' => 'ro',
	'required' => 1,
);

# Create DateTime object.
sub _datetime {
	my $self = shift;
	eval {
		require DateTime;
	};
	if ($EVAL_ERROR) {
		err "Cannot load 'DateTime' class.",
			'Error', $EVAL_ERROR;
	}
	my ($year, $month, $day) = split m/-/ms, $self->date;
	my ($hour, $min, $sec_mili) = split m/:/ms, $self->time;
	my ($sec, $mili) = split m/\./ms, $sec_mili;
	if (! defined $mili) {
		$mili = 0;
	}
	my $dt = eval {
		DateTime->new(
			'year' => $year,
			'month' => $month,
			'day' => $day,
			'hour' => $hour,
			'minute' => $min,
			'second' => $sec,
			'nanosecond' => $mili * 1000,
		);
	};
	if ($EVAL_ERROR) {
		err 'Cannot create DateTime object.',
			'Error', $EVAL_ERROR;
	}
	return $dt;
}

1;

__END__

=pod

=encoding utf8

=head1 NAME

Log::FreeSWITCH::Line::Data - Data object which represents FreeSWITCH log line.

=head1 SYNOPSIS

 use Log::FreeSWITCH::Line::Data;

 my $obj = Log::FreeSWITCH::Line::Data->new(%params);
 my $date = $obj->date;
 my $datetime_o = $obj->datetime_obj;
 my $file = $obj->file;
 my $file_line = $obj->file_line;
 my $message = $obj->message;
 my $raw = $obj->raw($raw);
 my $time = $obj->time;
 my $type = $obj->type;

=head1 METHODS

=head2 C<new>

 my $obj = Log::FreeSWITCH::Line::Data->new(%params);

Constructor.

Returns instance of object.

=over 8

=item * C<date>

 Date of log entry.
 Format of date is 'YYYY-MM-DD'.
 It is required.

=item * C<file>

 File in log entry.
 It is required.

=item * C<file_line>

 File line in log entry.
 It is required.

=item * C<message>

 Log message.

=item * C<raw>

 Raw FreeSWITCH log entry.

=item * C<time>

 Time of log entry.
 Format of time is 'HH:MM:SS'.
 It is required.

=item * C<type>

 Type of log entry.
 It is required.

=back

=head2 C<date>

 my $date = $obj->date;

Get log entry date.

Returns string with date in 'YYYY-MM-DD' format.

=head2 C<datetime_obj>

 my $datetime_o = $obj->datetime_obj;

Get DateTime object.

Returns DateTime object.

=head2 C<file>

 my $file = $obj->file;

Get file in log entry.

Returns string.

=head2 C<file_line>

 my $file_line = $obj->file_line;

Get file line in log entry.

Returns string.

=head2 C<message>

 my $message = $obj->message;

Get log message.

Returns string.

=head2 C<raw>

 my $raw = $obj->raw($raw);

Get or set raw FreeSWITCH log entry.

Returns string.

=head2 C<time>

 my $time = $obj->time;

Get log entry time.

Returns string with time in 'HH:MM:SS' format.

=head2 C<type>

 my $type = $obj->type;

Get log entry type.

Returns string.

=head1 ERRORS

 new():
         date required
         file required
         file_line required
         time required
         type required

 datetime_obj():
         Cannot create DateTime object.
                 Error: %s
         Cannot load 'DateTime' class.
                 Error: %s

=head1 EXAMPLE

=for comment filename=log_line_data_object.pl

 use strict;
 use warnings;

 use Log::FreeSWITCH::Line::Data;

 # Object.
 my $data_o = Log::FreeSWITCH::Line::Data->new(
         'date' => '2014-07-01',
         'file' => 'sofia.c',
         'file_line' => 4045,
         'message' => 'inbound-codec-prefs [PCMA]',
         'time' => '13:37:53.973562',
         'type' => 'DEBUG',
 );

 # Print out informations.
 print 'Date: '.$data_o->date."\n";

 # Output:
 # Date: 2014-07-01

=head1 DEPENDENCIES

L<DateTime>,
L<English>,
L<Error::Pure::Always>,
L<Error::Pure>,
L<Mo>.

=head1 SEE ALSO

=over

=item L<Log::FreeSWITCH::Line>

FreeSWITCH log line parsing and serializing.

=back

=head1 REPOSITORY

L<https://github.com/michal-josef-spacek/Log-FreeSWITCH-Line>

=head1 AUTHOR

Michal Josef Špaček L<mailto:skim@cpan.org>

L<http://skim.cz>

=head1 LICENSE AND COPYRIGHT

© 2014-2023 Michal Josef Špaček

BSD 2-Clause License

=head1 VERSION

0.09

=cut
